rfcs/20210731-tfjs-named-tensors.md

```diff
@@ -0,0 +1,557 @@
+* Have code more closely resemble the pseudocode and algorithms in papers.
+* Allow machine optimization of the ordering of indexes, transpositions and
+  permutations.
+
+## Motivation
+
+ML programming is currently much like assembly programming: you have to remember
+what each index in a tensor means, and refer to dimensions by index numbers, and
+when they change due to the operations you perform, you have to remember the new
+index position. Higher level abstractions could make this better and various
+ideas have been proposed around naming the dimensions in a Tensor e.g.
+http://nlp.seas.harvard.edu/NamedTensor
+
+The impact is likely largest for people who want to write ML algorithms, and for
+people using TFJS for education, e.g. to illustrate and visualize ML algorithms.
+However, it may also improve the experience for others using TFJS by providing
+better tool support, better auto-completion, and improved  optimization for
+evaluation.
+
+Here's an example TFJS implementation of single attention-head (For reference,
+see the [Tensorflow text tutorial on
+Transformers](https://www.tensorflow.org/text/tutorials/transformer) which
+provides a for walkthrough, explanation, and python code):
+
+```ts
+const queryM: = ...
+const keyM: = ...
+const valueM: = ...
+const xs: = ...
+
+// keys.shape == [seqLen, inputRep]
+const keys = tf.matMul(xs, keyM.read())
+
+// queries.shape == [seqLen, inputRep]
+const queries = tf.matMul(xs, queryM.read())
+
+// attention.shape == [seqLen (query), seqLen (key)]
+const attention = tf.softmax(
+ tf.div(tf.matMul(queries, keys, false, true), inputRepSizeSqrt)) as tf.Tensor2D;
+
+// values.shape == [seqLen, valueRepSize]
+const values = tf.matMul(xs, valueM.read())
+
+// attendedValues.shape == [seqLen (query), valueRepSize]
+const attendedValues = tf.matMul(attention, values);
+```
+
+The comments tell you which tensor index position corresponds to which dimension, and the author has to keep track of them and add appropriate comments to understand the code. It's hard to see the errors, especially without comments. For example, the following is very similar, will not typically result in a runtime error, but will give the wrong result:
+
+```ts
+const queryM: = ...
+const keyM: = ...
+const valueM: = ...
+const xs: = ...
+
+const keys = tf.matMul(xs, keyM.read())
+const queries = tf.matMul(xs, queryM.read())
+const attention = tf.softmax(
+ tf.div(tf.matMul(queries, keys, true, false), inputRepSizeSqrt)) as tf.Tensor2D;
+const values = tf.matMul(xs, valueM.read(), true, false)
+const attendedValues = tf.matMul(attention, values, false, true);
+```
+
+The proposed Named-Dimension library allows the same thing to be written as:
+
+```ts
+const queryM: GTensor<'inputRep', 'kqRep'> = ...
+const keyM: GTensor<'inputRep', 'kqRep'> = ...
+const valueM: GTensor<'inputRep', 'kqRep'> = ...
+const xs: GTensor<'inputRep', 'seqLen'> = ...
+
+const inputKeys = xs.inputRep.dot(keyM.inputRep).seqLen.rename('keySeqLen');
```

Comment:
In this example you first name it 'keys' (line 76) but here name it 'inputKeys'.  I would suggest keep everything consistent between the two (before and after) code chunks, except for what the RFC differences consist of.  Otherwise you are biasing the "readability" of the two code chunks.